instrgen/docs/design.md

```diff
@@ -0,0 +1,200 @@
+
+Go is perfect language for compile time instrumentation for two main reasons:
+- everything that's needed is in standard library
+- it's easy to control compilation process due to go compiler architecture
+
+## Design
+
+As mentioned above `instrgen` utilizes toolexec compiler switch (https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4058/).
+
+**NOTE** `driver` executable needs to be accessible via $PATH
+
+### Flow
+`instrgen` driver has two passes. First pass prepares some configuration for second pass based on provided settings,
+it means that the same process is executed twice in different context.
+
+It was designed and implemented this way to take input directly from user via command line switches and to do some
+initial analysis, however this can be spit into two separate tools or processes.
+
+`driver --inject  /testdata/basic yes main.main`
+
+Executable takes few parameters:
+
+- command name
+- file pattern - determines files taken into instrumentation
+- yes/no - determines whether code should be injected in-place
+- entry point - a function from which injection process start
+
+```sh
+usage driver --command [file pattern] replace entrypoint
+    command:
+	inject                                 (injects open telemetry calls into project code)
+	prune                                  (prune open telemetry calls
+```
+
+Below is the configuration generated by first pass and read by second one, when
+driver is executed as go compiler plugin.
+
+```json
+{
+"ProjectPath": ".",
+"FilePattern": "/testdata/basic",
+"Cmd": "inject",
+"Replace": "yes",
+"EntryPoint": {
+    "Pkg": "main",
+    "FunName": "main"
+ }
+}
+```
+
+First pass trigger the second one using following command:
+
+```sh
+go build -work -a -toolexec driver
+```
+
+### Instrumentation
+
+`instrgen` has modular architecture and to add instrumentation for new package, 2 steps are needed,
+implementation of specific `PackageRewriter` and its registration.
+
+```go
+// PackageRewriter interface does actual input package
+// rewriting according to specific criteria.
+type PackageRewriter interface {
+        // ID Dumps rewriter id.
+        Id() string
+        // Inject tells whether package should be rewritten.
+        Inject(pkg string, filepath string) bool
+        // ReplaceSource decides whether input sources should be replaced
+        // or all rewriting work should be done in temporary location.
+        ReplaceSource(pkg string, filePath string) bool
+        // Rewrite does actual package rewriting.
+        Rewrite(pkg string, file *ast.File, fset *token.FileSet, trace *os.File)
+        // WriteExtraFiles generate additional files that will be linked
+        // together to input package.
+        // Additional files have to be returned as array of file names.
+        WriteExtraFiles(pkg string, destPath string) []string
+}
+```
+
+
+#### Example rewriter
+
+```go
+// FunDumper rewrites all functions according to FilePattern.
```

Comment:
The example below could use more explanation.  FilePattern doesn't seem to be used.

Reply:
ok